conf/config_train.org
###############################################
# Parameters 
###############################################
IF_BATCH_J: 0
IF_W_BLOCKS: 0
# output different prompts for debugging purpose
VERBOSE: 0
# device to be used cuda or cpu
DEVICE: cuda
# float precision
FLOAT_PRECISION: 32

######################################
# Simulation Parameters
#####################################
# Number of batches (ie initializations)
N_BATCH: 1
# Time step in s
DT: 0.0025
# total simulation time in s
DURATION: 10
# time to start saving simulation files
T_STEADY: 1.0
# Saving to files every T_WINDOW
T_WINDOW: 0.1

##########################################
# Parameters for the stimulus presentation
##########################################
# stimulus as a cosine shape
TASK: 'None'
# # time for stimuli onset in s 
T_STIM_ON: [1.0, 3.5]
# # time for stimuli offset in s
T_STIM_OFF: [2.0, 4.5]
# stimuli strengths
I0: [0.2, 0.0, 0.0]
# stimuli phases
PHI0: [180.0, 90.0]
# stimuli phases
PHI1: [180.0, 90.0]
# stimuli tuning
SIGMA0: [1.0, 1.0, 1.0]

##########################################
# Network parameters
##########################################
# number of populations
N_POP: 2
# number of neurons
N_NEURON: 2500
# number of average presynaptic input (set to 1 if all to all)
K: 125.0
# fraction of neuron in each population (must sum to 1)
frac: [0.8, 0.2]

##########################################
# Voltage Dynamics for LIF model
##########################################
# Threshold in mV
V_THRESH: -50.0
# Resting potential in mV
V_REST: -70.0

##########################################
# Transfert Function for rate model
##########################################
# set to 0 for threshold linear, 1 for sigmoid
TF_TYPE: 'relu'
# threshold
THRESH: [15.0, 15.0]

##########################################
# Dynamics of the rates
##########################################
# set to 0 if instantaneous rates, 1 if exponentially decaying
RATE_DYN: 1
# rate time constants
TAU: [0.04, 0.02]

##########################################
# Dynamics of the recurrent inputs
##########################################
# set to 0 if instantaneous, 1 if exponentially decaying
SYN_DYN: 1
# Synaptic time constants for each population
TAU_SYN: [0.01, 0.01]

##########################################
# NMDA currents
##########################################
IF_NMDA: 1
R_NMDA: 1.0
TAU_NMDA: [0.5, 0.5]

##########################################
# Feedforward inputs
##########################################
LIVE_FF_UPDATE: 0
# External inputs strengths
Ja0: [2.0, 1.0]
# External rate
M0: 5.0
# Variance of the noise
VAR_FF: [0.0, 0.0]
# To add an attentional switch
# if BUMP_SWITCH[i] == 1 it sets Iext[i] to zero before stimulus presentation
BUMP_SWITCH: [1, 1]

##########################################
# Synaptic strengths           
##########################################
# Network's gain
GAIN: 1.0
# Synaptic strengths
Jab: [1., -1.5, 1, -1]

####################
# Plasticity
####################
# adds short term plasticity
IF_STP: 1
USE: 0.03
TAU_FAC: 1.0
TAU_REC: 0.2
J_STP: 10.0
STP_TYPE: 'markram'

##############
# Connectivity
##############
# seed for connectivity None or float
SEED: 1
# CON_TYPE can be 'all2all' or 'sparse'
CON_TYPE: 'sparse'
# PROBA_TYPE can be 'cosine', 'cosine_spec' or 'lr'
PROBA_TYPE: ['lr', 'None', 'None', 'None']

# strength of the asymmetries if all to all
SIGMA: [0.0, 0.0, 0.0, 0.0]
# tuning of the recurrent connections
KAPPA: [4.0, 0.0, 0.0, 0.0]

# phase of the connectivity
PHASE: 0.0

##########
# Low rank
##########
LR_EVAL_WIN: 0
LR_TRAIN: 0

# RANK: 2
# LR_MEAN: [0.0, 0.0]
# LR_COV: [[1.0, 0.0],[0.0, 1.0]]

# RANK: 3
# LR_MEAN: [0.0, 0.0, 0.0]
# LR_COV: [[1.0, 0.0, 0.0],[0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]

# RANK: 4
LR_MEAN: [0.0, 0.0, 0.0, 0.0]
LR_COV: [[1.0, 0.9, 0.0, 0.0],[0.9, 1.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.9], [0.0, 0.0, 0.9, 1.0]]


src/configuration.py
from yaml import safe_load
import numpy as np

import torch
from torch.distributions import MultivariateNormal

from src.utils import set_seed

class Configuration():

    def __init__(self, conf_name, repo_root):
        self.conf_file = repo_root + '/conf/' + conf_name
            
    def forward(self, **kwargs):
        parameters = safe_load(open(self.conf_file, "r"))
        parameters.update(kwargs)
        
        self.__dict__.update(parameters)

        if self.FLOAT_PRECISION == 32:
            self.FLOAT = torch.float
        else:
            self.FLOAT = torch.float64
        
        self.device = torch.device(self.DEVICE)
        
        # Set seed for the connectivity/input vectors
        set_seed(self.SEED)

        # create networks' constants
        init_const(self)
        
        return self
    
    def __call__(self, **kwargs):
        return self.forward(**kwargs)


def init_const(model):
    ##########################################
    # creating time steps from continuous time
    ##########################################
    model.N_STEADY = int(model.T_STEADY / model.DT)
    model.N_WINDOW = int(model.T_WINDOW / model.DT)
    model.N_STEPS = int(model.DURATION / model.DT) + model.N_STEADY + model.N_WINDOW

    model.N_STIM_ON = np.array([int(i / model.DT) + model.N_STEADY for i in model.T_STIM_ON])
    model.N_STIM_OFF = [int(i / model.DT) + model.N_STEADY for i in model.T_STIM_OFF]
    
    ##########################################
    # defining N and K per population
    ##########################################
    model.Na = []
    model.Ka = []
    
    # K is set to 1 in all2all nets
    if 'all2all' in model.CON_TYPE:
        model.K = 1.0

    for i_pop in range(model.N_POP):
        model.Na.append(int(model.N_NEURON * model.frac[i_pop]))
        # model.Ka.append(model.K * const.frac[i_pop])
        model.Ka.append(model.K)

    model.Na = torch.tensor(model.Na, dtype=torch.int, device=model.device)
    model.Ka = torch.tensor(model.Ka, dtype=model.FLOAT, device=model.device)
    model.csumNa = torch.cat((torch.tensor([0], device=model.device), torch.cumsum(model.Na, dim=0)))

    # slices[i] contains neurons' idx for population i
    model.slices = []
    for i_pop in range(model.N_POP):
        model.slices.append(slice(model.csumNa[i_pop], model.csumNa[i_pop + 1]))

    if model.VERBOSE:
        print("Na", model.Na, "Ka", model.Ka, "csumNa", model.csumNa)

    ##########################################
    # defining integration constants
    ##########################################

    # rate dynamics
    model.TAU = torch.tensor(model.TAU, dtype=model.FLOAT, device=model.device)
    model.EXP_DT_TAU = torch.ones(model.N_NEURON, dtype=model.FLOAT, device=model.device)
    model.DT_TAU = torch.ones(model.N_NEURON, dtype=model.FLOAT, device=model.device)

    for i_pop in range(model.N_POP):
        model.EXP_DT_TAU[model.slices[i_pop]] = torch.exp(-model.DT / model.TAU[i_pop])
        model.DT_TAU[model.slices[i_pop]] = model.DT / model.TAU[i_pop]

    # synaptic dynamics
    model.TAU_SYN = torch.tensor(model.TAU_SYN, dtype=model.FLOAT, device=model.device)
    model.EXP_DT_TAU_SYN = torch.ones(model.N_NEURON, dtype=model.FLOAT, device=model.device)
    model.DT_TAU_SYN = torch.ones(model.N_NEURON, dtype=model.FLOAT, device=model.device)

    for i_pop in range(model.N_POP):
        model.EXP_DT_TAU_SYN[model.slices[i_pop]] = torch.exp(-model.DT / model.TAU_SYN[i_pop])
        model.DT_TAU_SYN[model.slices[i_pop]] = model.DT / model.TAU_SYN[i_pop]

    # NMDA dynamics
    if model.IF_NMDA:
        model.TAU_NMDA = torch.tensor(model.TAU_NMDA, dtype=model.FLOAT, device=model.device)
        model.EXP_DT_TAU_NMDA = torch.ones(model.N_NEURON, dtype=model.FLOAT, device=model.device)
        model.DT_TAU_NMDA = torch.ones(model.N_NEURON, dtype=model.FLOAT, device=model.device)

        for i_pop in range(model.N_POP):
            model.EXP_DT_TAU_NMDA[model.slices[i_pop]] = torch.exp(-model.DT / model.TAU_NMDA[i_pop])
            model.DT_TAU_NMDA[model.slices[i_pop]] = model.DT / model.TAU_NMDA[i_pop]

    ##########################################
    # defining connectivity constants
    ##########################################
    model.PROBA_TYPE = np.array(model.PROBA_TYPE).reshape(model.N_POP, model.N_POP)
    model.SIGMA = torch.tensor(model.SIGMA, dtype=model.FLOAT, device=model.device).view(model.N_POP, model.N_POP)
    model.KAPPA = torch.tensor(model.KAPPA, dtype=model.FLOAT, device=model.device).view(model.N_POP, model.N_POP)
    model.PHASE = torch.tensor(model.PHASE * torch.pi / 180.0, dtype=model.FLOAT, device=model.device)
    
    model.PHI0 = torch.tensor(model.PHI0, dtype=model.FLOAT, device=model.device).unsqueeze(0) * torch.pi / 180.0
    # model.PHI1 = torch.tensor(model.PHI1, dtype=model.FLOAT, device=model.device).unsqueeze(0) * torch.pi / 180.0
    
    if 'dual' in model.TASK:
    # if 'lr' in model.PROBA_TYPE[0, 0]:
        mean_ = torch.tensor(model.LR_MEAN, dtype=model.FLOAT, device=model.device)
        cov_ = torch.tensor(model.LR_COV, dtype=model.FLOAT, device=model.device)

        if cov_[0, 0] == cov_[0, 1]:
            # print('Using Hopfield like low rank')

            mean_ = mean_[[0, 2]]
            cov_ = torch.tensor(([cov_[0,0], cov_[0, 2]], [cov_[2, 0], cov_[2,2]]),
                                dtype=model.FLOAT, device=model.device)

            multivariate_normal = MultivariateNormal(mean_, cov_)
            model.PHI0 = multivariate_normal.sample((model.Na[0],)).T
            model.PHI0 = torch.stack((model.PHI0[0], model.PHI0[0], model.PHI0[1], model.PHI0[1]))
        else:
            # print('Using Francesca like low rank')
            multivariate_normal = MultivariateNormal(mean_, cov_)
            model.PHI0 = multivariate_normal.sample((model.Na[0],)).T
        
        del mean_, cov_
        del multivariate_normal
